### PR TITLE
feature : added string, docker, dns functions and command

### PR DESCRIPTION
**Pull request by odebretseni**

### CODE REVIEW

File: tc/app/docker/container/restart [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  if [[ -n "$2" ]]; then
+    eval "${__key}"='$2'
+    shift 2
+  else
+    set "${__key}"
+    shift
+  fi
+done
+
+__cid=$(docker ps -q -f status=running -f name=^/tc-home-"$container_name")
```
Comment (BrunIF) [inline tc/app/docker/container/restart:21]: better filter value in quotes
```
__cid=$(docker ps -q -f "status=running" -f "name=^/tc-home-$container_name")
```

File: tc/app/docker/container/restart [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+if [ "$container_name" == "app" ]; then
+  container_name="home"
+fi
+if [ ! "${__cid}" ]; then
```
Comment (BrunIF) [inline tc/app/docker/container/restart:25]: it's the wrong condition
You get an error if a docker daemon not started.

File: tc/app/helpers/append_string_after [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+# shellcheck source=$CLI_DIR/tc/framework/functions
+. "${$CLI_DIR/tc/framework/functions}"
+
+# shellcheck source=$CLI_DIR/tc/app/dns/functions
+. "${$CLI_DIR/tc/app/dns/functions}"
```
Comment (BrunIF) [inline tc/app/helpers/append_string_after]: change

File: tc/app/helpers/delete_string [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+# shellcheck source=$CLI_DIR/tc/framework/functions
+. "${$CLI_DIR/tc/framework/functions}"
+
+# shellcheck source=$CLI_DIR/tc/app/dns/functions
+. "${$CLI_DIR/tc/app/dns/functions}"
```
Comment (BrunIF) [inline tc/app/helpers/delete_string]: change

File: tc/app/helpers/functions [hunk 5]
```diff
@@ -10,3 +10,124 @@ function _check_bin_file() {
     _message "$__MSG" "error" "$_error_code"
   fi
 }
+
+function _replace_string() {
+  __OS=$(tcctl helpers get_os)
```
Comment (BrunIF) [inline tc/app/helpers/functions:15]: inside tcctl command, we don't use tcctl. Only functions.

File: tc/app/helpers/functions [hunk 6]
```diff
@@ -10,3 +10,124 @@ function _check_bin_file() {
...
+    echo "Error -" "$__command"
+    exit 1
+  fi
+  if [[ "$_error_code" -eq "0" ]]; then
+    _message "String $1 was replacing in $3 file"
+  else
+    _message "$__MSG" "error" "$_error_code"
+    exit 1
+  fi
+}
```
Comment (BrunIF) [inline tc/app/helpers/functions:92]: duplicate with replace_string. Use parameters for functions.

File: tc/app/helpers/functions [hunk 7]
```diff
@@ -10,3 +10,124 @@ function _check_bin_file() {
...
+
+function _delete_string() {
+  __OS=$(uname -s)
+  __delete_string=$1
+  __file_to_edit=$2
+  if [[ $__OS == "Darwin" ]]; then
+    __command="gsed "
+  else
+    __command="sed "
+  fi
```
Comment (BrunIF) [inline tc/app/helpers/functions:102]: you can create function for this code. It's copy/paste.

File: tc/app/helpers/insert_string [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+# shellcheck source=$CLI_DIR/tc/framework/functions
+. "${$CLI_DIR/tc/framework/functions}"
+
+# shellcheck source=$CLI_DIR/tc/app/dns/functions
+. "${$CLI_DIR/tc/app/dns/functions}"
```
Comment (BrunIF) [inline tc/app/helpers/insert_string]: change

File: tc/app/helpers/insert_string.help [hunk 9]
```diff
@@ -0,0 +1,5 @@
+
+tcctl helpers insert_string --insert-string "STRING1" --file-name "FILE"
+
+--insert-string      "the string to insert" [REQUIRE]
+--file-name          "file name"  [REQUIRE]
```
Comment (BrunIF) [inline tc/app/helpers/insert_string.help:5]: Require:
--insert-string      "the string to insert"
--file-name          "file name" 

Optional:
--optianal_key

It's better

File: tc/app/helpers/replace_string [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+# shellcheck source=$CLI_DIR/tc/framework/functions
+. "${$CLI_DIR/tc/framework/functions}"
+
+# shellcheck source=$CLI_DIR/tc/app/dns/functions
+. "${$CLI_DIR/tc/app/dns/functions}"
```
Comment (BrunIF) [inline tc/app/helpers/replace_string]: change

File: tc/app/helpers/replace_string [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+done
+
+if [ -f "$file_name" ]; then
+  _replace_string "$find_string" "$replace_string" "$file_name"
+  __error_code=$?
+  if [[ $__error_code -gt "0" ]]; then
+    __MSG="File $file_name doesn't exists"
+    _message "$__MSG" "error" "$__error_code"
+  fi
+fi
```
Comment (BrunIF) [inline tc/app/helpers/replace_string:31]: same error

File: tc/app/helpers/replace_string.help [hunk 12]
```diff
@@ -0,0 +1,6 @@
+
+tcctl helpers replace_string --find-string "STRING1" --replace-string "STRING2" --file-name "FILE"
+
+--find-string        "the string to replace" [REQUIRE]
+--replace-string     "the string to which we replace" [REQUIRE]
+--file-name          "file name"  [REQUIRE]
```
Comment (BrunIF) [inline tc/app/helpers/replace_string.help:6]: change

File: tc/app/helpers/replace_string_with_escapes [hunk 13]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+# shellcheck source=$CLI_DIR/tc/framework/functions
+. "${$CLI_DIR/tc/framework/functions}"
+
+# shellcheck source=$CLI_DIR/tc/app/dns/functions
+. "${$CLI_DIR/tc/app/dns/functions}"
```
Comment (BrunIF) [inline tc/app/helpers/replace_string_with_escapes]: change

File: tc/app/helpers/replace_string_with_escapes [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+done
+
+if [ -f "$file_name" ]; then
+  _replace_string_with_escapes "$find_string" "$replace_string" "$file_name"
+  __error_code=$?
+  if [[ $__error_code -gt "0" ]]; then
+    __MSG="File $file_name doesn't exists"
+    _message "$__MSG" "error" "$__error_code"
+  fi
+fi
```
Comment (BrunIF) [inline tc/app/helpers/replace_string_with_escapes:31]: change

File: tc/app/helpers/replace_string_with_escapes.help [hunk 15]
```diff
@@ -0,0 +1,6 @@
+
+tcctl helpers replace_string_with_escapes --find-string "STRING1" --replace-string "STRING2" --file-name "FILE"
+
+--find-string        "the string to replace" [REQUIRE]
+--replace-string     "the string to which we replace" [REQUIRE]
+--file-name          "file name" [REQUIRE]
```
Comment (BrunIF) [inline tc/app/helpers/replace_string_with_escapes.help:6]: change

**Review by BrunIF — CHANGES_REQUESTED** [review]

**Review by BrunIF — CHANGES_REQUESTED** [review]

File: tc/app/dns/cert_gen.help [hunk 16]
```diff
@@ -0,0 +1,18 @@
+
+tcctl dns cert_gen --dns-name "FILE NAME"
```
[the remark this answers is not in the document]
Comment (odebretseni) [reply tc/app/dns/cert_gen.help]: this is no longer there

**Review by BrunIF — APPROVED** [review]

**Review by BrunIF — APPROVED** [review]

**Review by vikfedkiv — APPROVED** [review]

**Pull request merged**